tests/utilities/discovery_service_process.py

```diff
@@ -0,0 +1,48 @@
+"""Class to create and terminate Discovery Service instance."""
+
+from types import TracebackType
+from typing import Optional, Type
+
+from typing_extensions import Self
+
+from ni_measurementlink_service._internal.discovery_client import (
+    _get_discovery_service_location,
+    _get_key_file_path,
+    _service_already_running,
+    _start_service,
+)
+
+
+class DiscoveryServiceProcess:
+    """Maintains the processes involved in creating and terminating discovery service."""
+
+    def __init__(self) -> None:
+        """Creates a DiscoveryServiceProcess instance."""
+        try:
+            self._proc = None
+            key_file_path = _get_key_file_path()
+            if _service_already_running(key_file_path):
+                return
+            self._proc = _start_service(_get_discovery_service_location(), key_file_path)
+        except Exception:
+            self._close_discovery_service()
+            raise
```

Comment:
@bkeryan The pipeline(Run Tests for all platforms, all acceptance tests) errors if I raise the exception here. There are 2 errors that occurs in `discovery_client.py`
1) Non windows platform - `NotImplementedError("Platform not supported")` error from `_get_key_file_directory()`
2) windows platform - FileNotFound error from `_get_discovery_service_location()`

`discovery_client.py` handles these exceptions by logging them.

Due to these errors, Discovery Service is not even started and the acceptance tests pass showing that these tests are independent of Discovery Service running or not

So, my question is 
1) Whether we should validate the loggers for appropriate error messages in acceptance tests? or
2) Since the tests are independent on Discovery Service, can we fake the discovery service stub?

Reply:
Another option is to have the fixture call pytest.skip if the discovery service is not installed, like nidaqmx-python does. This is pretty easy to do and would unblock this PR, but we still need to figure out our strategy for testing Python code that uses real MeasurementLink core services.

Reply:
@bkeryan we want some clarifications on the Acceptance tests. So as we know acceptance tests starts and tries to register the measurements to the discovery service, 
If Discovery Service  is successfully started, it is not terminated
If Discovery Service  doesn't start successfully, exceptions are logged and tests pass

In the second case, the tests passing even though the measurement is not even registered to the discovery service is conveying that the overall workflow is not tested in the "Acceptance Tests" and I assume that Acceptance tests are end to end testing that tests the complete workflow of running a measurement. Please correct me If I'm wrong here.

So, we want to make the discovery service up and running for windows (which is only supported) such that acceptance tests works with the discovery service. To achieve this we are proposing the following approaches:

**1) Make Acceptance tests run only for windows:**
Mark all acceptance tests as windows only and start and terminate the discovery service using the `discovery_service_process` fixture. Doing this we will be testing the complete workflow needed to run a measurement.
(or)
**2) Use Fake discovery client stub for non windows platforms:**
Run acceptance tests for all platforms and use a fake discovery service if it is a non windows platform and use `discovery_service_process` for windows.
(or)
**3) Add new acceptance tests to verify discovery logs:**
Add new acceptance tests to verify the logs of Discovery service, thus we can test the error logs for non windows platform and success logs for windows platform

Please share your views on the approaches and direct on what approach to take or any other different direction will be helpful.

Ps: If we are going for approach 1, there is an `FileNotFound` error in pipeline, we are unsure of why that's occurring. please help us in debugging it.

Reply:
> 1) Make Acceptance tests run only for windows:
Mark all acceptance tests as windows only and start and terminate the discovery service using the discovery_service_process fixture. Doing this we will be testing the complete workflow needed to run a measurement.

> Ps: If we are going for approach 1, there is an FileNotFound error in pipeline, we are unsure of why that's occurring. please help us in debugging it.

```
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Program Files\\National Instruments\\Shared\\MeasurementLink\\MeasurementLinkServices.json'
```

The GitHub-hosted runner does not have MeasurementLink installed, so the JSON registration file does not exist. Note that this is separate from the key file. The registration file is created by the MeasurementLink installer, and the key file is created by the discovery service itself.
 
For this PR, I think it is reasonable for the discovery service test fixture to call pytest.skip in these situations:
- sys.platform != "win32"
- JSON registration file does not exist

I think skipping in the test fixture is better than adding `@pytest.skipif(sys.platform != "win32", reason="...")` to a bunch of tests.

> 2) Use Fake discovery client stub for non windows platforms:
Run acceptance tests for all platforms and use a fake discovery service for a non windows platform and use discovery_service_process for windows.

This adds more non-Windows test coverage but it's more effort/complication than skipping the tests on non-Windows platforms.

> 3) Add new acceptance tests to verify discovery logs:
Add new acceptance tests to verify the logs of Discovery service, thus we can test the error logs for non windows platform and success logs for windows platform

I don't think this is worth it. I would rather that we improve the error handling so that it's not necessary to check the logs for errors.

> Please share your views on the approaches and direct on what approach to take or any other different direction will be helpful.

I like this approach best:
- Now: have discovery fixture skip on non-Windows
- Now: have discovery fixture skip when MeasurementLink is not installed (registration file not found)
- Separately: improve error handling to exit when registration fails
- Separately: add self-hosted GitHub runner w/MeasurementLink installed (note that we might have a simpler way to do this than what we did for nidaqmx-python)

@dixonjoel @pbirkhol-ni

Reply:
```
Now: have discovery fixture skip on non-Windows
Now: have discovery fixture skip when MeasurementLink is not installed (registration file not found)
```
Thanks @bkeryan for the approach. I have added if conditions in `discovery_service_process` fixture in `conftest.py` to skip tests for  non-Windows and when MeasurementLink is not installed (registration file not found) with appropriate messages.

As per the other items:
`Separately: improve error handling to exit when registration fails` - This will be addressed in the upcoming PR.

`Separately: add self-hosted GitHub runner w/MeasurementLink installed (note that we might have a simpler way to do this than what we did for nidaqmx-python)` - I have created a **[Technical Debt 2486477](https://dev.azure.com/ni/DevCentral/_workitems/edit/2486477): Add Self hosted GitHub runner to run tests in measurementlink-python pipeline** to track.

Reply:
We have 'spawningpool' that we use for running the LabVIEW tests: https://github.com/ni/measurementlink-labview/settings/actions/runners
It's already set up as a self-hosted runner. I wonder if it's feasible to use it on two repos.

Reply:
In the meantime, it seems like we should be running the skipped tests locally with MeasurementLink installed before submitting PRs.

Reply:
Skipping tests is sufficient for this PR.

Adding a self-hosted runner is a matter for another PR, and we probably need to create a work item for it too.